File: CSharpCodeBase/components/character/playercontroller.lua
local Class = require 'utils.hump.class'
local CharacterEntity = require 'entities.player.character'
local Component = require 'components.component'

local PlayerController = Class({__includes=Component, init = function (self, entity)      
      Component.init(self, entity)
      self.entity.playerController = self
      self.slots = {}      
      self.countSlots = 0
      self.maxCountSlots = GameController.database:Get("heroes", "Common/maxCountSlots")
      self.currentSlot = 0                  
      self.characterGear = luanet.GameFacade.characterGear
      self.characterGear:SetupFinalCallback(function (idx) self:OnUIFinalSelectionCallback(idx) end)
      self.characterGearEnabled = false
end})

function PlayerController:OnUIFinalSelectionCallback(idx)
  print("Select slot: " .. idx)
  if idx + 1 ~= currentSlot then
    self:SelectSlot(idx + 1)
  end
end

function PlayerController:LoadAndAdd(entityName)
  local characterEntity = self:LoadSlot(entityName)
  self:AddSlot(characterEntity)
end

function PlayerController:LoadSlot(entityName)      
  local bundleName = GameController.database:Get('heroes', entityName .. '/bundle')
  local character = BundleUtils.CreateFromBundle(bundleName)
  character.name = entityName
  Transform.AddComponent(character, "LuaMapper")
  local characterEntity = CharacterEntity(character)
  return characterEntity
end

function PlayerController:AddSlot(entity)
  if self.countSlots < self.maxCountSlots then        
    entity.gameObject:SetActive(false)
    table.insert(self.slots, entity)
    self.countSlots = self.countSlots + 1    
    return true    
  end
  return false
end

function PlayerController:DeleteSlot(number)
  if number > 1 and number <= self.countSlots then
    local deletedSlot = self.currentSlot
    self:SelectSlot (1)
    deletedSlot = table.remove(self.slots, deletedSlot)
    deletedSlot:Destroy ()
    self.countSlots = self.countSlots - 1
    return true
  end
  return false
end

function PlayerController:DeleteCurrentSlot()
  local deletedSlot = self.currentSlot
  self:SelectSlot (1)
  deletedSlot = table.remove(self.slots, deletedSlot)
  deletedSlot:Destroy ()
  self.countSlots = self.countSlots - 1
  return true
end


function PlayerController:GetSlot(number)
  if number > 0 and number <= self.countSlots then
    return self.slots[number]
  else
    return nil
  end
end

function PlayerController:SelectSlot(number)
  if number ~= self.currentSlot and number > 0 and number <= self.countSlots then
    if self.currentSlot ~= 0 then
      local oldSlot = self:GetCurrentSlot()
      oldSlot.gameObject:SetActive(false)
      oldSlot:OnDeselectCharacter()
      oldSlot:Message("OnDeselectCharacter")
      GameController:RemoveEntity(oldSlot)
      self.currentSlot = number
    else      
      self.currentSlot = 1
    end
    local newSlot = self:GetCurrentSlot()
    newSlot:SetPosition(self.entity:GetPosition())
    newSlot:SetRotation(self.entity:GetRotation())
    newSlot.gameObject:SetActive(true)
    GameController:AddEntity(newSlot)
    newSlot:OnSelectCharacter()
    newSlot:Message("OnSelectCharacter")
    GameController.camera:LoadForEntity(newSlot)
    return true
  else
    return false
  end
end

function PlayerController:GetCurrentSlot()
  if self.currentSlot > 0 then
    return self:GetSlot(self.currentSlot)
  else
    return nil
  end
end

function PlayerController:Update()
  Component.Update(self)
  local slot = self:GetCurrentSlot()
  if slot then
    self.entity:SetPosition(slot:GetPosition())
    self.entity:SetRotation(slot:GetRotation())
  end
  if not self.characterGearEnabled and GameController.inputService:LeftBumperIsPressed() then
    self.characterGearEnabled = true
    for i = 1,5 do 
      local islot = self.slots[i]
      if islot ~= nil then
        self.characterGear:SetupSlot(i - 1, islot.config:Get('cgear_icon_left'), 
          islot.config:Get('cgear_icon_top'),
          islot.config:Get('cgear_icon_bottom'))
      else
        self.characterGear:SetupSlot(i - 1, '', '', '')
      end
    end
    self.characterGearSelected = self.currentSlot - 1
    self.characterGear:Show(self.currentSlot - 1)
    GameController.inputService:PushFrame("character_gear")
  end
  if self.characterGearEnabled and not GameController.inputService:LeftBumperIsPressed("character_gear") then
    self.characterGearEnabled = false
    self.characterGear:Hide()
    GameController.inputService:PopFrame()
  end
  if self.characterGearEnabled then
    if GameController.inputService:IsGamepad() then
      local lookVector = GameController.inputService:GetLookValue("character_gear")
      local selected = self.characterGearSelected
      --selected = self.currentSlot - 1
      if lookVector:Length() > 0.4 then
        local angle = math.atan2(lookVector.z, lookVector.x) * 180 / math.pi
        while angle < 0 do angle = angle + 360 end
        if angle > 315 or angle <= 45 then
          selected = 1
        elseif angle > 225 and angle <= 315 then
          selected = 0
        elseif angle > 135 and angle <= 225 then
          selected = 3
        else 
          selected = 2
        end
      end
      if self.slots[selected + 1] == nil then
        selected = self.characterGearSelected
        --selected = self.currentSlot - 1
      end
      if selected ~= self.characterGearSelected then
        -- update
        self.characterGear:UpdateSelected(selected)
        self.characterGearSelected = selected
      end
    end
  end
end

return PlayerController

File: CSharpCodeBase/components/character/playercontroller.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using UnityEngine;
using System.Collections;
using MainGame.core;

namespace MainGame {
    public class PlayerController : MonoBehaviour {
        private int countSlots = 0;
        private int maxCountSlots = 1;
        private int currentSlot = 0;
        private int characterGearSelected = 0;
        private bool characterGearEnabled = false;
        private List<GameObject> slots = new List<GameObject>();
        private CharacterGear characterGear = GameFacade.characterGear;


        /*public void init (self, entity)      ;
              Component.init(self, entity);
              this.entity.playerController = self;
              this.slots = {}      
              this.countSlots = 0;
              this.maxCountSlots = GameController.database:Get("heroes", "Common/maxCountSlots");
              this.currentSlot = 0                  ;
              this.characterGear = luanet.GameFacade.characterGear;
              this.characterGear:SetupFinalCallback(function (idx) self:OnUIFinalSelectionCallback(idx) }){
              this.characterGearEnabled = false;
        }})
       */

        public void Awake() {
            this.maxCountSlots = GetComponent<config>.Get("heroes", "Common/maxCountSlots");
            this.characterGear.SetupFinalCallback(/*callvack*/);
        }

        public void OnUIFinalSelectionCallback(int idx) {
            if (idx + 1 != currentSlot) {
                SelectSlot(idx + 1);
            }
        }

        public GameObject GetSlot(int number) {
            if (number > 0 && number <= this.countSlots) {
                return this.slots[number];
            } else {
                return null;
            }
        }

        public GameObject GetCurrentSlot() {
            if (this.currentSlot > 0) {
                return GetSlot(this.currentSlot);
            } else {
                return null;
            }
        }

        //FIXME
        public bool SelectSlot(int number) {
            if (number != this.currentSlot && number > 0 && number <= this.countSlots) {
                if (this.currentSlot != 0) {
                    var oldSlot = GetCurrentSlot();
                    oldSlot.SetActive(false);
                    SendMessage("OnDeselectCharacter");
                    this.currentSlot = number;
                } else {
                    this.currentSlot = 1;
                }
                var newSlot = GetCurrentSlot();
                newSlot.transform.position = transform.position;
                newSlot.transform.rotation = transform.rotation;
                newSlot.SetActive(true);
                SendMessage("OnSelectCharacter");
                GameController.camera.LoadForEntity(newSlot);
                return true;
            } else {
                return false;
            }
        }
        //FIXME
        public GameObject LoadSlot(string entityName) {
            var bundleName = GameController.database.Get("heroes", entityName + "/bundle");
            var character = BundleUtils.CreateFromBundle(bundleName);
            character.name = entityName;
            var characterEntity = new CharacterEntity(character);
            return characterEntity;
        }

        public bool AddSlot(GameObject entity) {
            if (this.countSlots < this.maxCountSlots) {
                entity.SetActive(false);
                this.slots.Add(entity);
                this.countSlots = this.countSlots + 1;
                return true;
            }
            return false;
        }
        //FIXME
        public bool DeleteSlot(int number) {
            if (number > 1 && number <= this.countSlots) {
                var deletedSlot = this.currentSlot;
                SelectSlot(1);
                this.slots.RemoveAt(deletedSlot);
                //deletedSlot:Destroy ();
                this.countSlots = this.countSlots - 1;
                return true;
            }
            return false;
        }

        public void LoadAndAdd(string entityName) {
            var characterEntity = LoadSlot(entityName);
            AddSlot(characterEntity);
        }

        public bool DeleteCurrentSlot() {
            var deletedSlot = this.currentSlot;
            SelectSlot(1);
            this.slots.RemoveAt(deletedSlot);
            //deletedSlot: Destroy();
            this.countSlots = this.countSlots - 1;
            return true;
        }

        public void Update() {
            var slot = GetCurrentSlot();
            if (slot != null) {
                transform.position = slot.transform.position;
                transform.rotation = slot.transform.rotation;
            }
            if (!this.characterGearEnabled && GameController.inputService.LeftBumperIsPressed()) {
                this.characterGearEnabled = true;
                for (int i = 0; i < 4; i++) {
                    var islot = this.slots[i];
                    if (islot != null) {
                        this.characterGear.SetupSlot(i - 1, islot.config.Get("cgear_icon_left"),
                          islot.config.Get("cgear_icon_top"),
                          islot.config.Get("cgear_icon_bottom"));
                    } else {
                        this.characterGear.SetupSlot(i - 1, "", "", "");
                    }
                }
                this.characterGearSelected = this.currentSlot - 1;
                this.characterGear.Show(this.currentSlot - 1);
                GameController.inputService.PushFrame("character_gear");
            }
            if (this.characterGearEnabled && !GameController.inputService.LeftBumperIsPressed("character_gear")) {
                this.characterGearEnabled = false;
                this.characterGear.Hide();
                GameController.inputService.PopFrame();
            }
            if (this.characterGearEnabled) {
                if (GameController.inputService.IsGamepad()) {
                    var lookVector = GameController.inputService.GetLookValue("character_gear");
                    var selected = this.characterGearSelected;
                    //selected = this.currentSlot - 1
                    if (lookVector.magnitude > 0.4f) {
                        var angle = Mathf.Atan2(lookVector.z, lookVector.x) * 180 / Mathf.PI;
                        while (angle < 0) {
                            angle = angle + 360;
                        }
                        if (angle > 315 || angle <= 45) {
                            selected = 1;
                        } else if (angle > 225 && angle <= 315) {
                            selected = 0;
                        } else if (angle > 135 && angle <= 225) {
                            selected = 3;
                        } else {
                            selected = 2;
                        }
                    }
                    if (this.slots[selected + 1] == null) {
                        selected = this.characterGearSelected;
                        //selected = this.currentSlot - 1
                    }
                    if (selected != this.characterGearSelected) {
                        // update
                        this.characterGear.UpdateSelected(selected);
                        this.characterGearSelected = selected;
                    }
                }
            }
        }
    }
}